### PR TITLE
Update mGGA implementation to new API and add unrestricted case

### PR DESCRIPTION
**Pull request by dmejiar**
@wavefunction91 This adds restricted and unrestricted tau-mGGAs. The code for Laplacian mGGAs is also there but it is deactivated for now.  I marked it as WIP as I did not have time to test the code.

### CODE REVIEW

**Comment by dmejiar** [comment]
@wavefunction91 I have tested the restricted and unrestricted VXC and everything looks good. 

I have not tested gradients as our code is not yet ready for that. We can comment that out, your call.

**Comment by wavefunction91** [comment]
@dmejiar I added an issue in ExchCXX to track the laplacian check
https://github.com/wavefunction91/ExchCXX/issues/34

File: src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx
Comment (wavefunction91) [inline src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx:736]:
```suggestion
      if ( mmat_x and mmat_y and mmat_z ) {
```

File: src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx
Comment (wavefunction91) [inline src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx:745]:
```suggestion
      if ( mmat_x and mmat_y and mmat_z ) {
```

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp
Comment (wavefunction91) [inline src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp:139]: I know I'm not the best at this either - but it might be good to comment why these factors are here. Magic numbers are hard to maintain, especially since they often contain contributions from multiple places (e.g. derivatives, spin, etc)

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp
Comment (wavefunction91) [inline src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp:222]: This is the laplacian check right? I would actually store this somewhere and just set it to true (with a comment to update once the lapl requirement is checkable)

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp
Comment (wavefunction91) [inline src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_grad.hpp:168]: Until we can check this, I would probably opt to disable this code path - the last thing we need is having something be accessible just to find out it doesn't work. We can discuss offline if there's an easy way to check this implementation (compare with NWChem, etc)

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp
Comment (wavefunction91) [inline src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp:228]: Per previous comment, store this in a variable with the TODO so we can make sure it propagates once we add the logic

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp
Comment (wavefunction91) [inline src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp:331]: We might be able to hack `gau2grid` to quickly evaluate the laplacian as opposed to the full hessian. Would definitely cut down on memory and should make the evaluation a bit faster (50% max)

**Review by wavefunction91 — CHANGES_REQUESTED** [review]
Before checking this in, we'll need to add some unit tests. We can intercept ExaChem data directly if that's helpful. I have an example of how to write a fully compliant reference file here:
https://github.com/wavefunction91/GauXC/blob/a777fa23c996633ea6d2855ec060905ac6b6c615/tests/standalone_driver.cxx#L447-L486

Essentially - we need the basis, density, molecule, and EXC/VXC written out in that standard HDF5 file.

Also, we're going to need to add a check in the GPU code to preclude mGGAs for the time being. I'll add the implementation soon after this is merged.

File: src/xc_integrator/local_work_driver/host/local_host_work_driver.cxx
Comment (dmejiar) [reply src/xc_integrator/local_work_driver/host/local_host_work_driver.cxx:81]: That's right, but we need 9 out of the 10 components of the full 3rd derivative matrix to get the derivatives of the Laplacian:
- xxx + xyy + xzz
- xxy + yyy + yyz
- xxz + yyz + zzz

File: src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx
Comment (dmejiar) [reply src/xc_integrator/local_work_driver/host/reference_local_host_work_driver.cxx:123]: I did not notice that piece of code before. If the bfn/gradients are guaranteed to be contiguous in the CPU that way is certainly better.

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp
Comment (dmejiar) [reply src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp:331]: Agreed

File: src/xc_integrator/local_work_driver/host/local_host_work_driver.cxx
Comment (wavefunction91) [reply src/xc_integrator/local_work_driver/host/local_host_work_driver.cxx:81]: That's a good point. Disregard.

**Comment by dmejiar** [comment]
> Essentially - we need the basis, density, molecule, and EXC/VXC written out in that standard HDF5 file.

I have generated such files, but to enable the tests we need to either use the LibXC backend or have built-in implementations of mGGA functionals (tau and lapl) in ExchCXX. 
 
> Also, we're going to need to add a check in the GPU code to preclude mGGAs for the time being. I'll add the implementation soon after this is merged.

I added an early check while setting up the integral factory, do you think this would work?

**Comment by wavefunction91** [comment]
> LibXC our builtin mGGA implementations
OK - can you make a quick list of the mGGA functionals you'd like to see proper infrastructure for? Also can you add these reference files to the `tests/reference_data` directory?

> I added an early check
Yep, that works for now. Thanks!!

File: src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp
Comment (wavefunction91) [reply src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp:334]: Unless you want to add this, we can punt for a bit. I've added an issue (https://github.com/wavefunction91/GauXC/issues/85) to track

Comment (dmejiar) [reply src/xc_integrator/replicated/host/reference_replicated_xc_host_integrator_exc_vxc.hpp:334]: Let's punt it

**Comment by wavefunction91** [comment]
@dmejiar Any info about how the reference files were generated? Bfn tolerances? Grids? Pruning?

**Comment by dmejiar** [comment]
> @dmejiar Any info about how the reference files were generated? Bfn tolerances? Grids? Pruning?

Hopefully the names of the file convey some of that info. These were obtained with the UltraFineGrid, SSF partitioning, and Robust pruning. I am not sure at what bfn tolerances you refer (we don't modify them in ExaChem) so they must have been the default

**Comment by wavefunction91** [comment]
Can you share how you construct the XC functional for SCAN? Do you just add `MGGA_X_SCAN`, or is it compounded with `MGGA_C_SCAN`? 

Also, for the cytosine test cases, are the UKS different spin states? Or are they just RKS run though the UKS code? If the latter, I'd prefer an actual UKS test is added, the way the UKS code works is pretty uninteresting for RKS data.

**Comment by dmejiar** [comment]
`MGGA_X_SCAN` plus `MGGA_C_SCAN` for the tauMGGA cases and `MGGA_X_R2SCANL` plus `MGGA_C_R2SCANL` for the Laplacian MGGA cases. The UKS data was generated for the +1 species (doublet).

**Comment by wavefunction91** [comment]
Unit tests added, but one of the reference files is missing data
```
dbwy@shire:~/Documents/Development/GauXC/tests/ref_data$ h5dump -H cytosine_r2scanl_cc-pvdz_ufg_ssf_robust.hdf5 
HDF5 "cytosine_r2scanl_cc-pvdz_ufg_ssf_robust.hdf5" {
GROUP "/" {
   DATASET "BASIS" {
      DATATYPE  H5T_COMPOUND {
         H5T_STD_I32LE "NPRIM";
         H5T_STD_I32LE "L";
         H5T_STD_I32LE "PURE";
         H5T_ARRAY { [16] H5T_IEEE_F64LE } "ALPHA";
         H5T_ARRAY { [16] H5T_IEEE_F64LE } "COEFF";
         H5T_ARRAY { [3] H5T_IEEE_F64LE } "ORIGIN";
      }
      DATASPACE  SIMPLE { ( 63 ) / ( 63 ) }
   }
   DATASET "MOLECULE" {
      DATATYPE  H5T_COMPOUND {
         H5T_STD_I32LE "Atomic Number";
         H5T_IEEE_F64LE "X Coordinate";
         H5T_IEEE_F64LE "Y Coordinate";
         H5T_IEEE_F64LE "Z Coordinate";
      }
      DATASPACE  SIMPLE { ( 13 ) / ( 13 ) }
   }
}
}
```

I added the stubs for that unit test [here](https://github.com/dmejiar/GauXC/blob/4c7e1923ec1ac73651071c86340434a76edb65b8/tests/xc_integrator.cxx#L305-L313). Update the file, uncomment the test and we're g2g.

**Comment by wavefunction91** [comment]
@dmejiar Also, I just confirmed that this compiles / precludes the right paths when devices are enabled (even though there's a divergence with master after #87, I checked the merge locally and everything works as expected, no conflicts)

**Comment by dmejiar** [comment]
@wavefunction91  file updated

**Review by wavefunction91 — APPROVED** [review]
LGTM

**Pull request merged**
